### PR TITLE
Conn: send and rcv on error should close connection

### PR DESCRIPTION
**Pull request by Neofine**
Idea is based on letting sendRequest() function handle connection closing as it is waiting on a channel and we can use this channel to communicate with it by using errors.

### CODE REVIEW

**Comment by mmatczuk** [comment]
How about closing the connection on the spot?

**Comment by mmatczuk** [comment]
We have put a lot of work to make sure Close is safe, it's connected to pool, let's use it.
Let's also log something.

**Comment by Neofine** [comment]
> We have put a lot of work to make sure Close is safe, it's connected to pool, let's use it.

`Close()` is a method on `Conn` struct and we currently have no access to it from in neither of `ConnWriter` nor `connReader` so we would have to add `*Conn` to their structs or to the argument of their loops which both seem like not desirable solution.

> Let's also log something.

Great idea, thanks!

**Comment by mmatczuk** [comment]
If things go south it should auto close. Pool would replace it.

File: transport/conn.go [hunk 1]
```diff
@@ -424,9 +440,15 @@ func (c *Conn) Shard() int {
 	return int(c.shard)
 }
 
+func (c *Conn) deactivateConnClose() {
+	c.w.connClose = nil
+	c.r.connClose = nil
+}
+
```
Comment (mmatczuk) [inline transport/conn.go:447]: This is clearly a race.

Comment (mmatczuk) [reply transport/conn.go:447]: I suggest we use [sync.Once](https://pkg.go.dev/sync#example-Once) in Conn::Close to ensure the actual closing happens once.

File: transport/conn.go [hunk 2]
```diff
@@ -325,6 +340,7 @@ func WrapConn(conn net.Conn) (*Conn, error) {
 		},
 		metrics: m,
 	}
+	c.setConnClose()
```
Comment (mmatczuk) [inline transport/conn.go:343]: Please inline, and remove deactivateConnClose

File: transport/conn.go [hunk 3]
```diff
@@ -116,10 +120,11 @@ type connReader struct {
...
 
-	h      map[frame.StreamID]responseHandler
-	s      streamIDAllocator
-	closed bool
-	mu     sync.Mutex // mu guards h, s and closed
+	h         map[frame.StreamID]responseHandler
+	s         streamIDAllocator
+	closed    bool
+	mu        sync.Mutex // mu guards h, s and closed
+	connClose func()
```
Comment (mmatczuk) [inline transport/conn.go:127]: Not this group, add that before metrics.

File: transport/conn.go [hunk 4]
```diff
@@ -307,34 +318,42 @@ func OpenConn(addr string, localAddr *net.TCPAddr, cfg ConnConfig) (*Conn, error
...
 	m := new(connMetrics)
-	c := &Conn{
+	var c Conn
+	c = Conn{
 		conn: conn,
 		w: connWriter{
 			conn:      bufio.NewWriterSize(conn, ioBufferSize),
 			requestCh: make(chan request, requestChanSize),
 			metrics:   m,
+			connClose: c.closeOnce,
```
Comment (Neofine) [inline transport/conn.go:336]: To make this assign possible I had to declare `c` earlier in code that's why I have added
`var c Conn`, is it worth it?
Other option is just keep it as it was and add
```
c.w.connClose = c.connClose
c.r.connClose = c.connClose
```
after the `&Conn` declaration, what do you think?

**Comment by mmatczuk** [comment]
@Neofine I added 44bbb1c8d7e3d3c64d1f02ec88aee400b01f3416 to improve the closing.

Nice trick with var init before use.

**Pull request merged**